extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnSecurity.java

```diff
@@ -1,84 +1,418 @@
+                        FidoMDS3MetadataBLOBAsyncProvider.DEFAULT_BLOB_ENDPOINT, httpClient, trustAnchors);
+                something = new MetadataBLOBBasedTrustAnchorAsyncRepository(blobAsyncProvider);
+            }
+
+            return new WebAuthnAsyncManager(
+                    Arrays.asList(
+                            new NoneAttestationStatementAsyncVerifier(),
+                            new FIDOU2FAttestationStatementAsyncVerifier(),
+                            new PackedAttestationStatementAsyncVerifier(),
+                            new TPMAttestationStatementAsyncVerifier(),
+                            new AndroidKeyAttestationStatementAsyncVerifier(),
+                            new AndroidSafetyNetAttestationStatementAsyncVerifier(),
+                            new AppleAnonymousAttestationStatementAsyncVerifier()),
+                    new DefaultCertPathTrustworthinessAsyncVerifier(something),
+                    new DefaultSelfAttestationTrustworthinessAsyncVerifier(),
+                    objectConverter);
+
+        } else {
+            return WebAuthnAsyncManager.createNonStrictWebAuthnAsyncManager(objectConverter);
         }
-        if (config.transports().isPresent()) {
-            options.setTransports(config.transports().get());
+    }
+
+    private void addCert(KeyStore keyStore, String pemCertificate) throws CertificateException, KeyStoreException {
+        X509Certificate cert = JWS.parseX5c(pemCertificate);
+        CertInfo info = CertificateHelper.getCertInfo(cert);
+        keyStore.setCertificateEntry(info.subject("CN"), cert);
+    }
+
+    private static byte[] uUIDBytes(UUID uuid) {
+        Buffer buffer = Buffer.buffer(16);
+        buffer.setLong(0, uuid.getMostSignificantBits());
+        buffer.setLong(8, uuid.getLeastSignificantBits());
+        return buffer.getBytes();
+    }
+
+    public Uni<PublicKeyCredentialCreationOptions> getRegisterChallenge(String userName, String displayName,
+            RoutingContext ctx) {
+        if (userName == null || userName.isEmpty()) {
+            return Uni.createFrom().failure(new IllegalArgumentException("Username is required"));
         }
-        if (config.userVerification().isPresent()) {
-            options.setUserVerification(config.userVerification().get());
+        // default displayName to userName, but it's required really
+        if (displayName == null || displayName.isEmpty()) {
+            displayName = userName;
         }
-        webAuthn = WebAuthn.create(vertx, options)
-                // where to load/update authenticators data
-                .authenticatorFetcher(database::fetcher)
-                .authenticatorUpdater(database::updater);
-        origin = config.origin().orElse(null);
-        if (origin != null) {
-            Origin o = Origin.parse(origin);
-            domain = o.host();
+        String finalDisplayName = displayName;
+        String challenge = randomBase64URLBuffer();
```

Comment:
With the current implementation of the "webauthn/login-challenge" endpoint, a new challenge value is generated every time the endpoint is called. 

For this reason, the "webauthn/login-challenge" endpoint is implemented using a POST method, and its name focuses on "challenge" to highlight its side effects.
However, generating a new challenge value on each call is inappropriate. The challenge should use the value stored in the session if it exists, and only generate a new one if no value is present in the session.

In the current implementation, if the "webauthn/login-challenge" endpoint is called multiple times, only the challenge value from the last call is valid.
However, there are cases where only the last challenge being valid is problematic — for example, in passkeys autofill. 
With passkey autofill, `navigator.credentials.get` must be called on page load, which also requires fetching the challenge on page load.
https://passkeys.dev/docs/use-cases/bootstrapping/#authenticating-the-user

But what happens if multiple tabs with the autofill-enabled page are open? The autofill will work only on the last opened tab, while the others encounter an error due to a mismatch in the challenge. 
Therefore, instead of generating a new challenge value each time the endpoint is called, same challenge value should be kept in a session.

Reply:
I see… but the challenge is stored in the session. It's just re-created every time. I could make it re-use the challenge if it already exists, but the challenge is tied to the user name. How does passkeys autofill know which uesrname to invoke on page load?

Reply:
The session should not be tied to the user name.
In the current implementation, `getLoginChallenge` requires `userName` as a required parameter, but it shouldn't be mandatory. 
With [discoverable credentials](https://www.w3.org/TR/webauthn-3/#discoverable-credential), login is possible without specifying a list of credential IDs in [`allowCredentials`](https://www.w3.org/TR/webauthn-3/#dom-publickeycredentialrequestoptions-allowcredentials), so userName input is not required.

Reply:
Ah.
I'm not sure I fully understand. I find this spec fairly confusing, and I'm unsure about how user names or display names are related to webauthn.
So, you're saying a user name is not required for login. I suppose in terms of implementation we'll look up the credentials by credential ID, so fine. In terms of UI, though, are we supposed to make the user name even optional or there's no point in asking for a user name for logins?
And, we still require a user name for registration, correct?

Reply:
> So, you're saying a user name is not required for login. 

No, user name is not required for login.

For example, when you try to login GitHub with passkeys, you will see login page like this:
![image](https://github.com/user-attachments/assets/3b616a31-7d7d-4bd4-a4ae-1a0bb752713a)

You can login with passkeys without filling the username field.

> we still require a user name for registration, correct?

yes, user name or display name is used as a label for the credential in the account selector dialog.
When you click "Sign in with passkey" button, it will show dialog like this.
![image](https://github.com/user-attachments/assets/358e97c4-0542-4b3b-aa8c-0074452b3738)

The "ynojima" is a value specified as a "display name" at the point of registration.

When you create a webauthn credential with `requireResidentKey` true, a discoverable credential data is created in the security key (if you are using security key) or in the passkey manager (if you are using synced passkey manager like Google Password Manager), and the discoverable credential data is consisted from its credential id, private key, user name, display name, counter value and etc.
The user name and display name of the credential data is only used for the account selector dialog.

Reply:
If I do that, though, then I can't support my yubikey because it's not a passkey one, whereas requiring a username made it usable :-/
Not sure how to properly handle the UI for that case :-/

Reply:
Does "my YubiKey is not a passkey one" mean the a Yubikey that does not support ["Discoverable Credentials"](https://www.w3.org/TR/webauthn-3/#client-side-discoverable-public-key-credential-source)?
For versions of Yubikey that lack "Discoverable Credential" support, known as ["Non-Discoverable Credentials"](https://www.w3.org/TR/webauthn-3/#non-discoverable-credential), the `allowCredentials` option must be specified when calling `navigator.credentials.get`.

Thus, to accommodate Yubikeys that lack "Discoverable Credentials" support, the login form should begin by asking users to input their username and password as a first step to avoid the account enumeration issue.
Then, `navigator.credentials.get` need to be called with the `allowCredentials` option in a second step for two-factor authentication. 
Consequently, the endpoint that returns the "options" should include `allowCredentials` option for authenticated requests but omit it for unauthenticated requests.

Or, do you concern that users of Yubikeys without Discoverable Credential support might mistakenly attempt to use a "Sign in with Passkeys" button, which only supports Discoverable Credentials causing errors and a less satisfactory user experience? 
If so, instead of providing a "Sign in with Passkeys" button on the login screen, you could implement [Passkey Autofill](https://passkeys.dev/docs/use-cases/bootstrapping/) to display a dropdown only when passkeys are available. This approach can introduce passkeys smoothly without confusing users who lack passkey support.

Reply:
OK, well, yes. I have a Yubikey that does not support Discoverable Credentials.

I've always been able to use it to login, by providing a username and not any password. Requiring that I type a password before I use it defeats the purpose I think.

But I can't use it as a passkey.

So, on my browser, which supports passkeys, I have both options available: passkeys or non-discoverable credentials.

During account creation, it's the same flow: enter username, create credentials and the browser will ask me whether to create a passkeys or use my yubikey and get a non-discoverable credential. So no UI problem.

On login though, it looks like I need two flows: one with username -> yubikey, and another for passkeys (no username). That seems a bit confusing.

Reply:
When authenticating with only a username and a security key supporting "non-discoverable credentials," the "webauthn/login-challenge" endpoint needs to return `allowCredentials` data based on the username for unauthenticated requests. 
This could lead to account enumeration issue; however, prioritizing user-experience, it could be reasonable to return `allowCredentials` to unauthenticated requests. 
For Quarkus, how about offering a configuration option that allows framework users to choose whether to return `allowCredentials` for unauthenticated requests?

> On login though, it looks like I need two flows: one with username -> yubikey, and another for passkeys (no username). That seems a bit confusing.

Maybe to avoid this confusion, Google has designed its login form to request the username even for "Discoverable Credentials" (passkeys). 
It seems like everyone is still figuring out the best approach.

Reply:
> For Quarkus, how about offering a configuration option that allows framework users to choose whether to return `allowCredentials` for unauthenticated requests?

That makes a lot of sense.

> It seems like everyone is still figuring out the best approach.

Yes, every time it looks like the paint is still fresh 😅

Reply:
Actually, Google requests the username but won't let me use my yubikey without typing a password first. They're listed separately from passkeys, as "2FA keys" and they require a password first.

Reply:
GitHub has a fairly complex UI which also ATM denies using my yubikey as a login method, only allowing passkeys.
BUT there's a UI to "Upgrade your security key registration to a passkey", which asks me to register a PIN code with it because I can use it as a passkey.
Passkeys are listed differently to "security keys".
I assume the difference is that these sorts of keys are not 2FA alone, so they need to be paired with either a password or a PIN code 🤷

Reply:
So, turns out once I set a PIN on this security key, I can store up to 25 passkeys on it. This works on github, but not on my Quarkus app, so I must be missing some setting. I'll try to find the difference.

Reply:
WebAuthn was initially introduced as a non-discoverable credential on various sites, because Chrome had not supported discoverable credentials for a long time. At that time, it was labeled as "Security Key" or "2FA Keys".
Later, when Chrome began supporting discoverable credentials, these sites added support for discoverable credential as well, distinguishing them with the label "passkeys".

If you added your YubiKey to GitHub a few years ago, it was probably registered as a "2FA key" because "discoverable credentials" (passkeys) were not supported at the time.

> This works on github, but not on my Quarkus app, so I must be missing some setting. I'll try to find the difference.

As current Quarkus WebAuthn implementation requests WebAuthn credential with `ResidentKey.DISCOURAGED` option by default, Yubikey and other FIDO2 compliant security keys return non-discoverable credential (non-resident key is equal to non-discoverable credential). 
https://github.com/quarkusio/quarkus/blob/370c50829579f7409a9e91bb0a3cb6a94cc7110c/extensions/security-webauthn/runtime/src/main/java/io/quarkus/security/webauthn/WebAuthnSecurity.java#L234
If you specify `quarkus.webauthn.resident-key=required` or `quarkus.webauthn.resident-key=preferred`, I think you can create discoverable credential (resident key) in your Yubikey with your Quarkus app.
On a related note, Passkeys by Google Password Manager returns discoverable credential even if `ResidentKey.DISCOURAGED` is specified.

Reply:
> If you added your YubiKey to GitHub a few years ago, it was probably registered as a "2FA key" because "discoverable credentials" (passkeys) were not supported at the time.

Ah, that explains it.

> As current Quarkus WebAuthn implementation requests WebAuthn credential with ResidentKey.DISCOURAGED option by default, Yubikey and other FIDO2 compliant security keys return non-discoverable credential (non-resident key is equal to non-discoverable credential).

Should I change that default, then?

I was able to fix my problem indeed by changing it. I don't understand why, before to changing that setting, it let me use it for registration but not login. That option only appears to come into play for registration.

Reply:
With the default `ResidentKey.DISCOURAGED` option, a non-discoverable credential will be created at registration.
I assume in your local environment, the `login-challenge` endpoint currently does not return `allowCredentials` option.
If the `allowCredentials` option is not provided, non-discoverable credentials cannot be found, leading to the login failure.

You might wonder why authentication with a non-discoverable credential is not possible without the `allowCredentials` option.
The answer lies in the tricky implementation of non-discoverable credentials. Due to storage limitation of security keys, the authentication key is encrypted and stored within the credential ID, making it essential to provide the credential ID through the `allowCredentials` option. 
With this technique, hardware security keys like Yubikey can support unlimited number of non-discoverable credentials.
For more information, please refer to the page below: 
https://crypto.stackexchange.com/questions/105942/how-do-non-resident-keys-work-in-webauthn


> Should I change that default, then?

If you expect the created WebAuthn credential to function only as a second factor, `ResidentKey.DISCOURAGED` would be appropriate.
If you need a Discoverable Credential(passkey) that allows login without even requiring username input, then `ResidentKey.REQUIRED` is suitable.
As for the default value in Quarkus, it depends on which use case you want to prioritize, but if you choose `ResidentKey.REQUIRED`, please note that this would be a breaking change, as the previous Quarkus Security WebAuthn defaulted `quarkus.webauthn.require-resident-key` to false.

Reply:
Well, this PR is full of lots of breaking changes. I suppose the main use-case nowadays would be passkeys…

Reply:
Yes, if breaking change is allowed, I prefer `ResidentKey.REQUIRED` too :)

Reply:
For passkey use cases, it is preferable for the default value of `quarkus.webauthn.user-verification` to be set to `REQUIRED` too.

Reply:
I can do that.